SOF-108-Node-ADC-Function

ShaneTWilliams opened the pull request:
The code for the ADC is under `OnPod/Node/Examples/ADC`.

The current example contains an ADC class. When it is instantiated you must pass a 16-bit number where each bit represents an ADC channel (starting with channel 0 on the right). The channel's bit should be `1` if the channel is used, `0` if it isn't. The current example reads channels 5, 11 and 13, so `iActiveChannels` is `0b001010000010000`.

To boot and configure the ADC, it's `init` function should be called.

To read values from the active channel, call the `readChannels()` method. It updates the ADC object's `iADCData` attribute, a 16-integer array where the integer at any index `i` is the latest available data from channel `i`.

`ADC` is a macro in the SAMD core so I called the class `ExtADC` and the object intance `adc`. This can change. I also re-formatted the `variant.h` and `variant.cpp` files. @stuckatmarine let me know your thoughts on that.

Commit:
Initial commit

Commit:
Initial commit

Commit:
Update README.md

Commit:
Merge branch 'SOF-108-Node-ADC-Function' of https://github.com/ParadigmHyperloop/comp4-software into SOF-108-Node-ADC-Function

Commit:
Add custom SPI settings

Commit:
Fix syntax

Commit:
Reformat board variant files

Commit:
Initial commit

Commit:
Initial commit

Commit:
Add comments

Commit:
Fix comments

Commit:
Fix .gitignore

EdHurtig commented on OnPod/Node/Examples/ADC/adc.cpp:
```diff
@@ -0,0 +1,47 @@
+#include <SPI.h>
+#include "adc.h"
+
+ExtADC::ExtADC(uint16_t iActiveChannels) : iActiveChannels(iActiveChannels) {
```
Traditionally the name of a SPI driver reflects the model number/device class of the device the driver interacts with: In this case an ADS7953. I'd recommend naming this as such with the initializer taking the appropriate chip select pin and other SPI specific settings (but not initializing the SPI bus in the constructor). This will also make it easier when the node board is rev'd in ways that change pin mappings.

Also, having enableChannel/disableChannel helpers would be much cleaner IMO than passing a 16 bit bitmask to a constructor.

EdHurtig commented on OnPod/Node/Examples/ADC/main.cpp:
```diff
@@ -0,0 +1,16 @@
+#include "adc.h"
+
+// start ADC object ussing channels 5, 11, and 13
+ExtADC adc(0b0010100000100000);
```
See aforementioned comment about enableChannel/disableChannel a bitmask like this will be hard to keep track of

Commit:
Change name of ADC class from ExtADC to ADS7953

Commit:
Add functions `enableChannel` and `disableChannel`

Commit:
Add function `readSingleChannel`

ShaneTWilliams commented:
- Changed `ExtADC` to `ADS7953`
- Added methods `enableChannel` and `disableChannel`
- The SPIClass and SPISettings objects are now created before the ADC object and are passed to it's constructor as arguments
- As per @stuckatmarine 's suggestion, I added the `readSingleChannel` method to read a single channel once and return it's value

`main.cpp` now contains an example where the ADC is instantiated, channel 5 is enabled, then it is read in both single read and auto read mode
